3 - Assignment/240521 - Rust Basic + Roguelike Game, Week 4/binary-tree/src/lib.rs

```diff
@@ -20,6 +20,65 @@ pub struct BinaryTree<T: Ord> {
 
 // Implement `new`, `insert`, `len`, and `has`.
 
+impl<T: Ord> Subtree<T> {
+    pub fn len(&self) -> i32 {
+        if self.0.is_none() {
+            return 0;
+        }
+        let node = self.0.as_ref().unwrap();
+        node.left.len() + node.right.len() + 1
+    }
+
+    pub fn insert(&mut self, value: T) {
+        if self.0.is_none() {
+            self.0 = Some(Box::new(Node {
+                value: value,
+                left: Subtree(None),
+                right: Subtree(None),
+            }));
+            return;
+        }
+        let node = self.0.as_mut().unwrap();
+        match value.cmp(&node.value) {
+            std::cmp::Ordering::Less => node.left.insert(value),
+            std::cmp::Ordering::Greater => node.right.insert(value),
+            std::cmp::Ordering::Equal => {}
+        }
+    }
+
+    pub fn has(&self, value: &T) -> bool {
+        if self.0.is_none() {
+            return false;
+        }
+        let node = self.0.as_ref().unwrap();
+        match value.cmp(&node.value) {
+            std::cmp::Ordering::Less => node.left.has(value),
+            std::cmp::Ordering::Greater => node.right.has(value),
+            std::cmp::Ordering::Equal => true,
+        }
+    }
```

Comment:
여기도 패턴 매칭을 사용하면 좋습니다.
```suggestion
        match &self.0 {
            Some(n) => match value.cmp(&n.value) {
                Ordering::Less => n.left.has(value),
                Ordering::Equal => true,
                Ordering::Greater => n.right.has(value),
            },
            None => false,
        }
```